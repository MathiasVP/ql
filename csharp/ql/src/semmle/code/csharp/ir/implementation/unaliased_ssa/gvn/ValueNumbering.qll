private import internal.ValueNumberingInternal
private import internal.ValueNumberingImports

/**
 * Provides additional information about value numbering in IR dumps.
 */
class ValueNumberPropertyProvider extends IRPropertyProvider {
  override string getInstructionProperty(Instruction instr, string key) {
    exists(ValueNumber vn |
      vn = valueNumber(instr) and
      key = "valnum" and
      if strictcount(vn.getAnInstruction()) > 1 then result = vn.toString() else result = "unique"
    )
  }
}

/**
 * The value number assigned to a particular set of instructions that produce equivalent results.
 */
class ValueNumber extends TValueNumber {
  final string toString() { result = "GVN" }

  final string getDebugString() {
    result = "ValueNumber: " +
        strictconcat(this.getAnInstruction().getUnconvertedResultExpression().toString(), ", ")
  }

  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    this
        .getAnInstruction()
        .getLocation()
        .hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
  }

  /**
   * Gets the instructions that have been assigned this value number. This will always produce at
   * least one result.
   */
  final Instruction getAnInstruction() { this = valueNumber(result) }

  /**
   * Gets one of the instructions that was assigned this value number. The chosen instuction is
   * deterministic but arbitrary. Intended for use only in debugging.
   */
  final Instruction getExampleInstruction() {
    result = min(Instruction instr |
        instr = getAnInstruction()
      |
        instr order by instr.getBlock().getDisplayIndex(), instr.getDisplayIndexInBlock()
      )
  }

  /**
   * Gets an `Operand` whose definition is exact and has this value number.
   */
  final Operand getAUse() { this = valueNumber(result.getDef()) }

  final string getKind() {
    this instanceof TVariableAddressValueNumber and result = "VariableAddress"
    or
    this instanceof TInitializeParameterValueNumber and result = "InitializeParameter"
    or
    this instanceof TInitializeThisValueNumber and result = "InitializeThis"
    or
    this instanceof TStringConstantValueNumber and result = "StringConstant"
    or
    this instanceof TFieldAddressValueNumber and result = "FieldAddress"
    or
    this instanceof TBinaryValueNumber and result = "Binary"
    or
    this instanceof TPointerArithmeticValueNumber and result = "PointerArithmetic"
    or
    this instanceof TUnaryValueNumber and result = "Unary"
    or
    this instanceof TInheritanceConversionValueNumber and result = "InheritanceConversionr"
    or
    this instanceof TUniqueValueNumber and result = "Unique"
  }

  Language::Expr getAnExpr() { result = getAnInstruction().getUnconvertedResultExpression() }
}

/**
 * Gets the value number assigned to `instr`, if any. Returns at most one result.
 */
ValueNumber valueNumber(Instruction instr) { result = tvalueNumber(instr) }

/**
 * Gets the value number assigned to the exact definition of `op`, if any.
 * Returns at most one result.
 */
ValueNumber valueNumberOfOperand(Operand op) { result = valueNumber(op.getDef()) }
