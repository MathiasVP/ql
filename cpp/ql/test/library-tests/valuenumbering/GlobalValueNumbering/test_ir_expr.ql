import cpp
import semmle.code.cpp.valuenumbering.GlobalValueNumbering
import semmle.code.cpp.ir.ValueNumbering
import semmle.code.cpp.ir.IR

Expr ir(Expr e) {
  exists(Instruction i |
    e = i.getUnconvertedResultExpression() and
    result = valueNumber(i).getAnExpr()
  )
}

Expr ast(Expr e) { result = globalValueNumber(e).getAnExpr() }

from Expr e, Expr evn
where evn = ast(e) and not evn = ir(e)
select e, evn
