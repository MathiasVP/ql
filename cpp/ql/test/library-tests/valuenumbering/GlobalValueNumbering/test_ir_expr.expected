| test.cpp:5:3:5:13 | ... = ... | test.cpp:5:3:5:13 | ... = ... |
| test.cpp:6:3:6:13 | ... = ... | test.cpp:6:3:6:13 | ... = ... |
| test.cpp:7:3:7:7 | ... = ... | test.cpp:7:3:7:7 | ... = ... |
| test.cpp:10:16:10:16 | 1 | test.cpp:10:16:10:16 | 1 |
| test.cpp:16:3:16:24 | ... = ... | test.cpp:16:3:16:24 | ... = ... |
| test.cpp:16:7:16:24 | ... + ... | test.cpp:17:7:17:24 | ... + ... |
| test.cpp:16:7:16:24 | ... + ... | test.cpp:18:7:18:7 | x |
| test.cpp:16:17:16:24 | global01 | test.cpp:17:17:17:24 | global01 |
| test.cpp:17:3:17:24 | ... = ... | test.cpp:17:3:17:24 | ... = ... |
| test.cpp:17:7:17:24 | ... + ... | test.cpp:16:7:16:24 | ... + ... |
| test.cpp:17:17:17:24 | global01 | test.cpp:16:17:16:24 | global01 |
| test.cpp:18:3:18:7 | ... = ... | test.cpp:18:3:18:7 | ... = ... |
| test.cpp:18:7:18:7 | x | test.cpp:16:7:16:24 | ... + ... |
| test.cpp:21:16:21:16 | 2 | test.cpp:21:16:21:16 | 2 |
| test.cpp:29:3:29:24 | ... = ... | test.cpp:29:3:29:24 | ... = ... |
| test.cpp:30:3:30:17 | call to change_global02 | test.cpp:30:3:30:17 | call to change_global02 |
| test.cpp:31:3:31:24 | ... = ... | test.cpp:31:3:31:24 | ... = ... |
| test.cpp:32:3:32:7 | ... = ... | test.cpp:32:3:32:7 | ... = ... |
| test.cpp:35:16:35:16 | 3 | test.cpp:35:16:35:16 | 3 |
| test.cpp:43:3:43:24 | ... = ... | test.cpp:43:3:43:24 | ... = ... |
| test.cpp:44:3:44:9 | ... = ... | test.cpp:44:3:44:9 | ... = ... |
| test.cpp:44:9:44:9 | 0 | test.cpp:51:25:51:25 | 0 |
| test.cpp:44:9:44:9 | 0 | test.cpp:53:18:53:21 | (int)... |
| test.cpp:44:9:44:9 | 0 | test.cpp:56:39:56:42 | (int)... |
| test.cpp:44:9:44:9 | 0 | test.cpp:59:17:59:20 | (int)... |
| test.cpp:44:9:44:9 | 0 | test.cpp:88:12:88:12 | 0 |
| test.cpp:45:3:45:24 | ... = ... | test.cpp:45:3:45:24 | ... = ... |
| test.cpp:46:3:46:7 | ... = ... | test.cpp:46:3:46:7 | ... = ... |
| test.cpp:51:25:51:25 | 0 | test.cpp:44:9:44:9 | 0 |
| test.cpp:51:25:51:25 | 0 | test.cpp:53:18:53:21 | (int)... |
| test.cpp:51:25:51:25 | 0 | test.cpp:56:39:56:42 | (int)... |
| test.cpp:51:25:51:25 | 0 | test.cpp:59:17:59:20 | (int)... |
| test.cpp:51:25:51:25 | 0 | test.cpp:88:12:88:12 | 0 |
| test.cpp:51:25:51:25 | (unsigned int)... | test.cpp:51:25:51:25 | (unsigned int)... |
| test.cpp:53:10:53:13 | (int)... | test.cpp:53:10:53:13 | (int)... |
| test.cpp:53:10:53:13 | (int)... | test.cpp:56:21:56:24 | (int)... |
| test.cpp:53:10:53:13 | * ... | test.cpp:56:21:56:24 | * ... |
| test.cpp:53:18:53:21 | (int)... | test.cpp:44:9:44:9 | 0 |
| test.cpp:53:18:53:21 | (int)... | test.cpp:51:25:51:25 | 0 |
| test.cpp:53:18:53:21 | (int)... | test.cpp:53:18:53:21 | (int)... |
| test.cpp:53:18:53:21 | (int)... | test.cpp:56:39:56:42 | (int)... |
| test.cpp:53:18:53:21 | (int)... | test.cpp:59:17:59:20 | (int)... |
| test.cpp:53:18:53:21 | (int)... | test.cpp:88:12:88:12 | 0 |
| test.cpp:55:5:55:15 | ... = ... | test.cpp:55:5:55:15 | ... = ... |
| test.cpp:56:12:56:25 | (...) | test.cpp:56:12:56:25 | (...) |
| test.cpp:56:12:56:43 | ... && ... | test.cpp:56:12:56:43 | ... && ... |
| test.cpp:56:13:56:16 | (int)... | test.cpp:56:13:56:16 | (int)... |
| test.cpp:56:13:56:16 | (int)... | test.cpp:56:31:56:34 | (int)... |
| test.cpp:56:13:56:16 | (int)... | test.cpp:59:9:59:12 | (int)... |
| test.cpp:56:13:56:16 | * ... | test.cpp:56:31:56:34 | * ... |
| test.cpp:56:13:56:16 | * ... | test.cpp:59:9:59:12 | * ... |
| test.cpp:56:14:56:16 | ptr | test.cpp:56:47:56:49 | ptr |
| test.cpp:56:21:56:24 | (int)... | test.cpp:53:10:53:13 | (int)... |
| test.cpp:56:21:56:24 | (int)... | test.cpp:56:21:56:24 | (int)... |
| test.cpp:56:21:56:24 | * ... | test.cpp:53:10:53:13 | * ... |
| test.cpp:56:30:56:43 | (...) | test.cpp:56:30:56:43 | (...) |
| test.cpp:56:31:56:34 | (int)... | test.cpp:56:13:56:16 | (int)... |
| test.cpp:56:31:56:34 | (int)... | test.cpp:56:31:56:34 | (int)... |
| test.cpp:56:31:56:34 | (int)... | test.cpp:59:9:59:12 | (int)... |
| test.cpp:56:31:56:34 | * ... | test.cpp:56:13:56:16 | * ... |
| test.cpp:56:31:56:34 | * ... | test.cpp:59:9:59:12 | * ... |
| test.cpp:56:32:56:34 | ptr | test.cpp:56:47:56:49 | ptr |
| test.cpp:56:39:56:42 | (int)... | test.cpp:44:9:44:9 | 0 |
| test.cpp:56:39:56:42 | (int)... | test.cpp:51:25:51:25 | 0 |
| test.cpp:56:39:56:42 | (int)... | test.cpp:53:18:53:21 | (int)... |
| test.cpp:56:39:56:42 | (int)... | test.cpp:56:39:56:42 | (int)... |
| test.cpp:56:39:56:42 | (int)... | test.cpp:59:17:59:20 | (int)... |
| test.cpp:56:39:56:42 | (int)... | test.cpp:88:12:88:12 | 0 |
| test.cpp:56:47:56:49 | ptr | test.cpp:56:14:56:16 | ptr |
| test.cpp:56:47:56:49 | ptr | test.cpp:56:32:56:34 | ptr |
| test.cpp:56:47:56:49 | ptr | test.cpp:59:10:59:12 | ptr |
| test.cpp:59:9:59:12 | (int)... | test.cpp:56:13:56:16 | (int)... |
| test.cpp:59:9:59:12 | (int)... | test.cpp:56:31:56:34 | (int)... |
| test.cpp:59:9:59:12 | (int)... | test.cpp:59:9:59:12 | (int)... |
| test.cpp:59:9:59:12 | * ... | test.cpp:56:13:56:16 | * ... |
| test.cpp:59:9:59:12 | * ... | test.cpp:56:31:56:34 | * ... |
| test.cpp:59:10:59:12 | ptr | test.cpp:56:47:56:49 | ptr |
| test.cpp:59:17:59:20 | (int)... | test.cpp:44:9:44:9 | 0 |
| test.cpp:59:17:59:20 | (int)... | test.cpp:51:25:51:25 | 0 |
| test.cpp:59:17:59:20 | (int)... | test.cpp:53:18:53:21 | (int)... |
| test.cpp:59:17:59:20 | (int)... | test.cpp:56:39:56:42 | (int)... |
| test.cpp:59:17:59:20 | (int)... | test.cpp:59:17:59:20 | (int)... |
| test.cpp:59:17:59:20 | (int)... | test.cpp:88:12:88:12 | 0 |
| test.cpp:62:5:62:10 | result | test.cpp:65:10:65:15 | result |
| test.cpp:65:10:65:15 | result | test.cpp:62:5:62:10 | result |
| test.cpp:77:20:77:30 | (signed short)... | test.cpp:77:20:77:30 | (signed short)... |
| test.cpp:77:20:77:30 | (signed short)... | test.cpp:79:7:79:7 | v |
| test.cpp:79:7:79:7 | (int)... | test.cpp:79:7:79:7 | (int)... |
| test.cpp:79:7:79:7 | v | test.cpp:77:20:77:30 | (signed short)... |
| test.cpp:79:11:79:20 | (int)... | test.cpp:79:11:79:20 | (int)... |
| test.cpp:79:24:79:33 | (int)... | test.cpp:79:24:79:33 | (int)... |
| test.cpp:80:5:80:19 | ... = ... | test.cpp:80:5:80:19 | ... = ... |
| test.cpp:80:9:80:19 | (signed short)... | test.cpp:80:9:80:19 | (signed short)... |
| test.cpp:88:3:88:20 | ... = ... | test.cpp:88:3:88:20 | ... = ... |
| test.cpp:88:12:88:12 | 0 | test.cpp:44:9:44:9 | 0 |
| test.cpp:88:12:88:12 | 0 | test.cpp:51:25:51:25 | 0 |
| test.cpp:88:12:88:12 | 0 | test.cpp:53:18:53:21 | (int)... |
| test.cpp:88:12:88:12 | 0 | test.cpp:56:39:56:42 | (int)... |
| test.cpp:88:12:88:12 | 0 | test.cpp:59:17:59:20 | (int)... |
| test.cpp:88:12:88:12 | (void *)... | test.cpp:88:12:88:12 | (void *)... |
| test.cpp:105:11:105:12 | (Base *)... | test.cpp:105:11:105:12 | (Base *)... |
| test.cpp:105:11:105:12 | (Base *)... | test.cpp:106:14:106:35 | static_cast<Base *>... |
| test.cpp:105:11:105:12 | (Base *)... | test.cpp:107:11:107:12 | pb |
| test.cpp:105:15:105:15 | b | test.cpp:107:15:107:15 | b |
| test.cpp:105:15:105:15 | b | test.cpp:109:10:109:10 | y |
| test.cpp:106:14:106:35 | static_cast<Base *>... | test.cpp:105:11:105:12 | (Base *)... |
| test.cpp:106:14:106:35 | static_cast<Base *>... | test.cpp:106:14:106:35 | static_cast<Base *>... |
| test.cpp:106:14:106:35 | static_cast<Base *>... | test.cpp:107:11:107:12 | pb |
| test.cpp:107:11:107:12 | pb | test.cpp:105:11:105:12 | (Base *)... |
| test.cpp:107:11:107:12 | pb | test.cpp:106:14:106:35 | static_cast<Base *>... |
| test.cpp:107:15:107:15 | b | test.cpp:105:15:105:15 | b |
| test.cpp:109:10:109:10 | y | test.cpp:105:15:105:15 | b |
